File: py/src/tradekit/core/stats.py
"""Summaries of a set of closed trades.

Mirrors ``go/core/stats``. Every dashboard and every backtest report in the
projects tradekit replaces recomputed these same figures, and no two agreed on
all of them. Sharing the queries rather than the templates is the point: a win
rate should not depend on which project is displaying it.
"""

from __future__ import annotations

import datetime as dt
import math
from dataclasses import dataclass, field

from . import money
from .domain import ExitReason, Side, Trade
from .money import Money

__all__ = [
    "Metrics",
    "MixedModesError",
    "Point",
    "daily_returns",
    "equity_curve",
    "max_drawdown",
    "r_multiple",
    "sharpe",
    "summarize",
]


class MixedModesError(ValueError):
    """Paper and live trades were summarised together.

    An error rather than a silent merge because every aggregate -- win rate,
    profit factor, drawdown, the equity curve -- is meaningless when simulated
    fills are blended with real ones, and the resulting number looks entirely
    plausible.
    """


@dataclass(slots=True)
class Metrics:
    """Summary of a set of closed trades.

    Monetary fields are in minor units; ratios are plain fractions.
    """

    trades: int = 0
    wins: int = 0
    losses: int = 0
    breakeven: int = 0

    gross_profit: Money = money.ZERO
    gross_loss: Money = money.ZERO
    net_pnl: Money = money.ZERO
    charges: Money = money.ZERO

    win_rate: float = 0.0
    profit_factor: float = 0.0
    expectancy: Money = money.ZERO

    avg_win: Money = money.ZERO
    avg_loss: Money = money.ZERO

    max_drawdown: Money = money.ZERO
    max_drawdown_pct: float = 0.0

    avg_holding: dt.timedelta = dt.timedelta(0)

    by_exit_reason: dict[ExitReason, int] = field(default_factory=dict)


def summarize(trades: list[Trade]) -> Metrics:
    """Compute the metrics for a set of closed trades.

    Trades need not be sorted; they are ordered by exit time internally, because
    drawdown is meaningless over an arbitrary ordering.

    Raises:
        MixedModesError: If paper and live trades are mixed.

    """
    m = Metrics()
    if not trades:
        return m
    if len({t.paper for t in trades}) > 1:
        raise MixedModesError("stats: paper and live trades cannot be summarised together")

    ordered = sorted(trades, key=lambda t: t.exit_at)
    holding = dt.timedelta(0)
    gross_profit = gross_loss = net = charges = 0

    for t in ordered:
        m.trades += 1
        net += int(t.net_pnl)
        charges += int(t.charges)
        m.by_exit_reason[t.exit_reason] = m.by_exit_reason.get(t.exit_reason, 0) + 1
        holding += t.exit_at - t.entry_at
        if t.net_pnl > 0:
            m.wins += 1
            gross_profit += int(t.net_pnl)
        elif t.net_pnl < 0:
            m.losses += 1
            gross_loss += -int(t.net_pnl)
        else:
            m.breakeven += 1

    m.gross_profit, m.gross_loss = Money(gross_profit), Money(gross_loss)
    m.net_pnl, m.charges = Money(net), Money(charges)
    m.win_rate = m.wins / m.trades
    m.expectancy = Money(net // m.trades)
    m.avg_holding = holding / m.trades

    if m.wins:
        m.avg_win = Money(gross_profit // m.wins)
    if m.losses:
        m.avg_loss = Money(gross_loss // m.losses)

    if gross_loss > 0:
        m.profit_factor = gross_profit / gross_loss
    elif gross_profit > 0:
        # No losing trades. Infinity is the honest answer; a caller formatting
        # it must decide how to display that rather than being handed a
        # fabricated finite number.
        m.profit_factor = math.inf

    m.max_drawdown, m.max_drawdown_pct = max_drawdown(equity_curve(ordered, money.ZERO))
    return m


@dataclass(frozen=True, slots=True)
class Point:
    """One sample of the running equity curve."""

    at: dt.datetime
    equity: Money


def equity_curve(trades: list[Trade], opening: Money = money.ZERO) -> list[Point]:
    """Accumulate net P&L over trades in exit order, starting from ``opening``."""
    running = int(opening)
    curve: list[Point] = []
    for t in sorted(trades, key=lambda t: t.exit_at):
        running += int(t.net_pnl)
        curve.append(Point(t.exit_at, Money(running)))
    return curve


def max_drawdown(curve: list[Point]) -> tuple[Money, float]:
    """Largest peak-to-trough fall, as an amount and a fraction of that peak.

    The fraction is relative to each peak rather than to the starting equity, so
    a 10% fall late in a doubled account is reported as 10% and not as 5%.
    """
    if not curve:
        return money.ZERO, 0.0
    peak = int(curve[0].equity)
    worst, worst_pct = 0, 0.0
    for p in curve:
        peak = max(peak, int(p.equity))
        fall = peak - int(p.equity)
        if fall > worst:
            worst = fall
            worst_pct = fall / peak if peak > 0 else 0.0
    return Money(worst), worst_pct


def r_multiple(trade: Trade, initial_stop: Money) -> float | None:
    """A trade's result as a multiple of the risk taken on it.

    ``initial_stop`` is the stop in force at entry, not wherever it was trailed
    to. A trade risking 1 and making 3 is a 3R trade regardless of account size,
    which is what makes R comparable across instruments and across time.

    Returns ``None`` when the initial risk was zero, since no multiple exists.
    """
    risk = abs(int(trade.entry_price) - int(initial_stop))
    if risk == 0:
        return None
    move = int(trade.exit_price) - int(trade.entry_price)
    if trade.side is Side.SELL:
        move = -move
    return move / risk


def daily_returns(curve: list[Point], tz: dt.tzinfo | None = None) -> list[float]:
    """Aggregate an equity curve into per-day fractional returns.

    This is the input :func:`sharpe` needs.
    """
    if len(curve) < 2:
        return []

    days: list[tuple[str, int]] = []
    for p in curve:
        at = p.at.astimezone(tz) if tz else p.at
        key = at.date().isoformat()
        if days and days[-1][0] == key:
            days[-1] = (key, int(p.equity))
        else:
            days.append((key, int(p.equity)))

    out: list[float] = []
    for i in range(1, len(days)):
        prev = days[i - 1][1]
        if prev == 0:
            continue
        out.append((days[i][1] - prev) / abs(prev))
    return out


def sharpe(returns: list[float], periods_per_year: float = 252.0, risk_free: float = 0.0) -> float | None:
    """Annualised Sharpe ratio of a series of periodic returns.

    ``periods_per_year`` scales the result: 252 for daily returns on an equity
    calendar. ``risk_free`` is the per-period rate, not the annual one.

    Returns ``None`` when the series has no variance, since the ratio is
    undefined rather than infinite in any useful sense.
    """
    if len(returns) < 2 or periods_per_year <= 0:
        return None
    excess = [r - risk_free for r in returns]
    mean = sum(excess) / len(excess)
    # Sample standard deviation: the returns are a sample of the strategy's
    # behaviour, not the whole population of it.
    variance = sum((e - mean) ** 2 for e in excess) / (len(excess) - 1)
    if variance <= 0:
        return None
    return mean / math.sqrt(variance) * math.sqrt(periods_per_year)

File: go/core/stats/stats.go
// Package stats summarises a set of closed trades.
//
// Every dashboard and every backtest report in the projects tradekit replaces
// recomputed these same figures, and no two agreed on all of them. Sharing the
// queries rather than the templates is the point: a win rate should not depend
// on which project is displaying it.
package stats

import (
	"errors"
	"math"
	"sort"
	"time"

	"github.com/althk/tradekit/go/core/domain"
	"github.com/althk/tradekit/go/core/money"
)

// ErrMixedModes reports that paper and live trades were summarised together.
//
// It is an error rather than a silent merge because every aggregate below —
// win rate, profit factor, drawdown, the equity curve — is meaningless when
// simulated fills are blended with real ones, and the resulting number looks
// entirely plausible.
var ErrMixedModes = errors.New("stats: paper and live trades cannot be summarised together")

// Metrics is the summary of a set of closed trades. Monetary fields are in
// minor units; ratios are plain fractions, not percentages.
type Metrics struct {
	Trades    int
	Wins      int
	Losses    int
	Breakeven int

	GrossProfit money.Money // sum of winning trades' net P&L
	GrossLoss   money.Money // sum of losing trades' net P&L, as a positive amount
	NetPnL      money.Money
	Charges     money.Money

	WinRate      float64     // wins / trades
	ProfitFactor float64     // gross profit / gross loss; +Inf when there are no losses
	Expectancy   money.Money // average net P&L per trade

	AvgWin  money.Money
	AvgLoss money.Money // a positive amount

	MaxDrawdown    money.Money // largest peak-to-trough fall in the equity curve
	MaxDrawdownPct float64     // as a fraction of the peak it fell from

	AvgHolding time.Duration

	ByExitReason map[domain.ExitReason]int
}

// Summarize computes the metrics for a set of closed trades.
//
// Trades need not be sorted; they are ordered by exit time internally, because
// drawdown is meaningless over an arbitrary ordering.
func Summarize(trades []domain.Trade) (Metrics, error) {
	m := Metrics{ByExitReason: map[domain.ExitReason]int{}}
	if len(trades) == 0 {
		return m, nil
	}
	if mixedModes(trades) {
		return m, ErrMixedModes
	}

	ordered := make([]domain.Trade, len(trades))
	copy(ordered, trades)
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].ExitAt.Before(ordered[j].ExitAt) })

	var holding time.Duration
	for _, t := range ordered {
		m.Trades++
		m.NetPnL += t.NetPnL
		m.Charges += t.Charges
		m.ByExitReason[t.ExitReason]++
		holding += t.ExitAt.Sub(t.EntryAt)

		switch {
		case t.NetPnL > 0:
			m.Wins++
			m.GrossProfit += t.NetPnL
		case t.NetPnL < 0:
			m.Losses++
			m.GrossLoss += -t.NetPnL
		default:
			m.Breakeven++
		}
	}

	m.WinRate = float64(m.Wins) / float64(m.Trades)
	m.Expectancy = m.NetPnL / money.Money(m.Trades)
	m.AvgHolding = holding / time.Duration(m.Trades)

	if m.Wins > 0 {
		m.AvgWin = m.GrossProfit / money.Money(m.Wins)
	}
	if m.Losses > 0 {
		m.AvgLoss = m.GrossLoss / money.Money(m.Losses)
	}

	switch {
	case m.GrossLoss > 0:
		m.ProfitFactor = float64(m.GrossProfit) / float64(m.GrossLoss)
	case m.GrossProfit > 0:
		// No losing trades. Infinity is the honest answer; a caller
		// formatting it must decide how to display that rather than
		// being handed a fabricated finite number.
		m.ProfitFactor = math.Inf(1)
	}

	m.MaxDrawdown, m.MaxDrawdownPct = MaxDrawdown(EquityCurve(ordered, 0))
	return m, nil
}

func mixedModes(trades []domain.Trade) bool {
	first := trades[0].Paper
	for _, t := range trades[1:] {
		if t.Paper != first {
			return true
		}
	}
	return false
}

// Point is one sample of the running equity curve.
type Point struct {
	At     time.Time
	Equity money.Money
}

// EquityCurve accumulates net P&L over trades in exit order, starting from
// opening.
func EquityCurve(trades []domain.Trade, opening money.Money) []Point {
	ordered := make([]domain.Trade, len(trades))
	copy(ordered, trades)
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].ExitAt.Before(ordered[j].ExitAt) })

	curve := make([]Point, 0, len(ordered))
	running := opening
	for _, t := range ordered {
		running += t.NetPnL
		curve = append(curve, Point{At: t.ExitAt, Equity: running})
	}
	return curve
}

// MaxDrawdown returns the largest peak-to-trough fall in a curve, both as an
// amount and as a fraction of the peak it fell from.
//
// The fraction is relative to each peak rather than to the starting equity, so
// a 10% fall late in a doubled account is reported as 10% and not as 5%.
func MaxDrawdown(curve []Point) (money.Money, float64) {
	if len(curve) == 0 {
		return 0, 0
	}
	peak := curve[0].Equity
	var worst money.Money
	var worstPct float64
	for _, p := range curve {
		if p.Equity > peak {
			peak = p.Equity
		}
		if fall := peak - p.Equity; fall > worst {
			worst = fall
			if peak > 0 {
				worstPct = float64(fall) / float64(peak)
			}
		}
	}
	return worst, worstPct
}

// RMultiple expresses a trade's result as a multiple of the risk taken on it.
//
// initialStop is the stop in force at entry, not wherever it was trailed to.
// A trade risking 1 and making 3 is a 3R trade regardless of the account size,
// which is what makes R comparable across instruments and across time.
// It reports false when the initial risk was zero, since no multiple exists.
func RMultiple(t domain.Trade, initialStop money.Money) (float64, bool) {
	risk := (t.EntryPrice - initialStop).Abs()
	if risk == 0 {
		return 0, false
	}
	move := t.ExitPrice - t.EntryPrice
	if t.Side == domain.Sell {
		move = -move
	}
	return float64(move) / float64(risk), true
}

// DailyReturns aggregates an equity curve into per-day fractional returns,
// which is the input Sharpe needs.
func DailyReturns(curve []Point, loc *time.Location) []float64 {
	if len(curve) < 2 {
		return nil
	}
	if loc == nil {
		loc = time.UTC
	}

	type dayEnd struct {
		day    string
		equity money.Money
	}
	var days []dayEnd
	for _, p := range curve {
		key := p.At.In(loc).Format(time.DateOnly)
		if n := len(days); n > 0 && days[n-1].day == key {
			days[n-1].equity = p.Equity
			continue
		}
		days = append(days, dayEnd{day: key, equity: p.Equity})
	}

	returns := make([]float64, 0, len(days)-1)
	for i := 1; i < len(days); i++ {
		prev := days[i-1].equity
		if prev == 0 {
			continue
		}
		returns = append(returns, float64(days[i].equity-prev)/math.Abs(float64(prev)))
	}
	return returns
}

// Sharpe is the annualised Sharpe ratio of a series of periodic returns.
//
// periodsPerYear scales the result: 252 for daily returns on an equity
// calendar. riskFree is the per-period risk-free rate, not the annual one.
// It reports false when the series has no variance, since the ratio is
// undefined rather than infinite in any useful sense.
func Sharpe(returns []float64, periodsPerYear float64, riskFree float64) (float64, bool) {
	if len(returns) < 2 || periodsPerYear <= 0 {
		return 0, false
	}

	var sum float64
	for _, r := range returns {
		sum += r - riskFree
	}
	mean := sum / float64(len(returns))

	var sq float64
	for _, r := range returns {
		d := (r - riskFree) - mean
		sq += d * d
	}
	// Sample standard deviation: the returns are a sample of the strategy's
	// behaviour, not the whole population of it.
	variance := sq / float64(len(returns)-1)
	if variance <= 0 {
		return 0, false
	}
	return mean / math.Sqrt(variance) * math.Sqrt(periodsPerYear), true
}
